python-checks/src/main/java/org/sonar/python/checks/DjangoNonDictSerializationCheck.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.annotation.CheckForNull;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.symbols.Usage;
+import org.sonar.plugins.python.api.tree.Argument;
+import org.sonar.plugins.python.api.tree.AssignmentStatement;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6560")
+public class DjangoNonDictSerializationCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Use a dictionary object here, or set the \"safe\" flag to False.";
+  private static final String JSON_RESPONSE_FUNCTION_NAME = "django.http.JsonResponse";
+
+  private static final int MAX_RECURSION = 5;
+
+  @Override
+  public void initialize(Context context) {
+
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, ctx -> {
+      CallExpression callExpression = (CallExpression) ctx.syntaxNode();
+      Symbol symbol = callExpression.calleeSymbol();
+      if (symbol != null && JSON_RESPONSE_FUNCTION_NAME.equals(symbol.fullyQualifiedName())) {
+        checkForDictSerialization(ctx, callExpression);
+      }
+    });
+  }
+
+  private static void checkForDictSerialization(SubscriptionContext ctx, CallExpression callExpression) {
+    RegularArgument safe = TreeUtils.nthArgumentOrKeyword(2, "safe", callExpression.arguments());
+    if (safe == null || (safe.expression().is(Tree.Kind.NAME) && "True".equals(((Name) safe.expression()).name()))) {
+      RegularArgument dataArg = getFirstArgument(callExpression.arguments());
```

Comment:
Is the data argument positional only, or should we use `nthArgumentOrKeyword` here as well? (I didn't double check, this is a legitimate question)

Reply:
Exactly the data arg is positional only 
```
class JsonResponse(data, encoder=DjangoJSONEncoder, safe=True, json_dumps_params=None, **kwargs)
```

Reply:
I don't think this signature makes `data` positional only. It just means that there is no default value for the parameter (positional only arguments should be before a `/` delimiter argument, like: `def my_func(pos_only, /, keyword_or_pos, *, keyword_only): ...` - although it's true that some Typeshed stubs for builtins don't follow this convention).

In this case I could actually create a JsonResponse with the following call:
```
>>> response = JsonResponse(safe=False, data=42 )
>>> response
<JsonResponse status_code=200, "application/json">
```
So I think we should actually consider this order as being possible (granted it's very unlikely to happen in real cases).